File: 4_leetcode/C++/LC_1115_sem.cpp
#include <semaphore>
#include <iostream>
#include <thread>


class FooBar {
private:
    int n;
    counting_semaphore<1> sem_foo{1};
    counting_semaphore<1> sem_bar{0};

public:
    FooBar(int n) {
        this->n = n;
    }

    void foo(function<void()> printFoo) {
        
        for (int i = 0; i < n; i++) {
            sem_foo.acquire();
        	// printFoo() outputs "foo". Do not change or remove this line.
        	printFoo();
            sem_bar.release();
        }
    }

    void bar(function<void()> printBar) {
        
        for (int i = 0; i < n; i++) {
            sem_bar.acquire();
        	// printBar() outputs "bar". Do not change or remove this line.
        	printBar();
            sem_foo.release();
        }
    }
};

File: 4_leetcode/C#/LC_1115_sem.cs
using System;
using System.Threading;

public class FooBar {
    private int n;

    private SemaphoreSlim semFoo;
    private SemaphoreSlim semBar;

    public FooBar(int n) {
        this.n=n;

        semFoo=new SemaphoreSlim(1, 1);
        semBar=new SemaphoreSlim(0, 1);
    }

    public void Foo(Action printFoo) {
        
        for (int i = 0; i < n; i++) {
            semFoo.Wait();
        	// printFoo() outputs "foo". Do not change or remove this line.
        	printFoo();
            semBar.Release();
        }
    }

    public void Bar(Action printBar) {
        
        for (int i = 0; i < n; i++) {
            semBar.Wait();
            // printBar() outputs "bar". Do not change or remove this line.
        	printBar();
            semFoo.Release();
        }
    }
}